lib/chef/provider/user/mac.rb

```diff
@@ -0,0 +1,607 @@
+          dscl_set(user_plist, :shadow_hash, Plist.parse_xml(shadow_xml_plist))
+
+          user_plist
+        rescue Chef::Exceptions::PlistUtilCommandFailed, Chef::Exceptions::DsclCommandFailed
+          nil
+        end
+
+        # Map updated user information from the user_plist onto the current resource
+        # and instance variables.
+        def reload_with_user_plist
+          current_resource.uid(dscl_get(user_plist, :uid)[0])
+          current_resource.gid(dscl_get(user_plist, :gid)[0])
+          current_resource.home(dscl_get(user_plist, :home)[0])
+          current_resource.shell(dscl_get(user_plist, :shell)[0])
+          current_resource.comment(dscl_get(user_plist, :comment)[0])
+
+          shadow_hash = dscl_get(user_plist, :shadow_hash)
+          if shadow_hash
+            current_resource.password(shadow_hash[0]["SALTED-SHA512-PBKDF2"]["entropy"].string.unpack("H*")[0])
+            current_resource.salt(shadow_hash[0]["SALTED-SHA512-PBKDF2"]["salt"].string.unpack("H*")[0])
+            current_resource.iterations(shadow_hash[0]["SALTED-SHA512-PBKDF2"]["iterations"].to_i)
+          end
+
+          @auth_authority = dscl_get(user_plist, :auth_authority)
+          @guid = dscl_get(user_plist, :guid)
+          @admin_group_info = nil
+
+          current_resource.secure_token(secure_token_enabled?)
+          current_resource.admin(admin_user?)
+        end
+
+        #
+        # User Provider Callbacks
+        #
+
+        def create_user
+          cmd = [-"-addUser", new_resource.username]
+          cmd += ["-fullName", new_resource.comment]  if new_resource.property_is_set?(:comment)
+          cmd += ["-UID", new_resource.uid]           if new_resource.property_is_set?(:uid)
+          cmd += ["-shell", new_resource.shell]       if new_resource.property_is_set?(:shell)
+          cmd += ["-home", new_resource.home]         if new_resource.property_is_set?(:home)
+          cmd += ["-admin"]                           if new_resource.admin
+
+          # We can technically create a new user without the admin credentials
+          # but without them the user cannot enable SecureToken, thus they cannot
+          # create other secure users or enable FileVault full disk encryption.
+          if new_resource.property_is_set?(:admin_username) && new_resource.property_is_set?(:admin_password)
+            cmd += ["-adminUser", new_resource.admin_username]
+            cmd += ["-adminPassword", new_resource.admin_password]
+          end
+
+          # sysadminctl doesn't exit with a non-zero exit code if it encounters
+          # a problem. We'll check stderr and make sure we see that it finished
+          # correctly.
+          res = run_sysadminctl(cmd)
+          unless res.downcase =~ /creating user/
+            raise Chef::Exceptions::User, "error when creating user: #{res}"
+          end
+
+          # Wait for the user to be created and for dscl to flush everything
+          sleep 3
```

Comment:
if you can figure out a way to poll a bit more tightly as to when its really done instead of just a large sleep value that'd both be more correct under high load / low cpu situations and would be faster in the common case.